python/ql/test/experimental/dataflow/coverage/test.py

```diff
@@ -39,93 +42,241 @@ def test_floatnumber_literal():
+    n = await g.__anext__()
+    SINK(await g.athrow(TypeError)) # Flow missing
+
+def test_athrow():
+    runa(atest_athrow())
+
 # 6.3.1. Attribute references
+class C:
+    a = SOURCE
+
+def test_attribute_reference():
+    SINK(C.a) # Flow missing
+
+# overriding __getattr__ should be tested by the class coverage tests
+
 # 6.3.2. Subscriptions
+# This does not constitute dataflow (but could be taint flow)
+def example_subscription_string():
+    SINK("source"[0]) # Flow not expected
+
+def test_subscription_tuple():
+    SINK((SOURCE,)[0]) # Flow missing
+
+def test_subscription_list():
+    SINK([SOURCE][0]) # Flow missing
+
+def test_subscription_mapping():
+    SINK({"s":SOURCE}["s"]) # Flow missing
+
+# overriding __getitem__ should be tested by the class coverage tests
+
 # 6.3.3. Slicings
+l = [SOURCE]
+
+def test_slicing():
+    s = l[0:1:1]
+    SINK(s[0]) # Flow missing
+
+# The grammar seems to allow `l[0:1:1, 0:1]`, but the interpreter does not like it
+
 # 6.3.4. Calls
-# 6.4. Await expression
-# 6.5. The power operator
-# 6.6. Unary arithmetic and bitwise operations
-# 6.7. Binary arithmetic operations
-# 6.8. Shifting operations
-# 6.9. Binary bitwise operations
-# 6.10. Comparisons
-# 6.10.1. Value comparisons
-# 6.10.2. Membership test operations
-# 6.10.3. Identity comparisons
-# 6.11. Boolean operations
+def f(a, b):
+    return b
+
+def test_call_positional():
+    SINK(f(3, SOURCE))
+
+def test_call_keyword():
+    SINK(f(3, b=SOURCE)) # Flow missing
+
+def test_call_unpack_iterable():
+    SINK(f(3, *[SOURCE])) # Flow missing
+
+def test_call_unpack_mapping():
+    SINK(f(3, **{"b": SOURCE})) # Flow missing
+
+def f_extra_pos(a, *b):
+    return b[0]
+
+def test_call_extra_pos():
+    SINK(f_extra_pos(3, SOURCE)) # Flow missing
+
+def f_extra_keyword(a, **b):
+    return b["b"]
+
+def test_call_extra_keyword():
+    SINK(f_extra_keyword(3, b=SOURCE)) # Flow missing
+
+# return the name of the first extra keyword argument
+def f_extra_keyword_flow(**a):
+    return [*a][0]
+
+# call the function with our source as the name of the keyword arguemnt
+def test_call_extra_keyword_flow():
+    SINK(f_extra_keyword_flow(**{SOURCE: None})) # Flow missing
+
 # 6.12. Assignment expressions
+def test_assignment_expression():
+    x = 3
+    SINK(x := SOURCE) # Flow missing
+
 # 6.13. Conditional expressions
+def test_conditional_true():
+    SINK(SOURCE if True else 3) # Flow missing
+
+def test_conditional_false():
+    SINK(3 if False else SOURCE) # Flow missing
+
+# Condition is evaluated first, so x is SOURCE once chosen
+def test_conditional_evaluation_true():
+    x = 3
+    SINK(x if (SOURCE == (x := SOURCE)) else 3) # Flow missing
+
+# Condition is evaluated first, so x is SOURCE once chosen
+def test_conditional_evaluation_false():
+    x = 3
+    SINK(3 if (3 == (x := SOURCE)) else x) # Flow missing
+
 # 6.14. Lambdas
-# 6.15. Expression lists
-# 6.16. Evaluation order
-# 6.17. Operator precedence
+def test_lambda():
```

Comment:
You could probably expand these tests a bit - for example lambdas with multiple arguments, `NONSOURCE` test case, parameter packs etc.

Reply:
So basically a suite mirroring the one for function calls?

Reply:
Done